File: code/291/solution.cpp
#include <iostream>
#include <vector>
using namespace std;
const long long upper = 5e15;
vector<long long> v;
int main() {
    // 2n^2 + 2n + 1;
    v.push_back(0);
    for (long long n = 1; n < 5e7; n++) {
        long long now = 2 * n * n + 2 * n + 1;
        v.push_back(now);
    }

    int ans = 0;
    for (long long i = 1; i < v.size(); i++) {
        long long p = v[i];
        if (p == 2 * i * i + 2 * i + 1)
            ans++;
        if (p <= 1)
            continue;
        for (long long j = i; j < v.size(); j += p) {
            while (v[j] % p == 0)
                v[j] /= p;
        }
        for (long long j = p - 1 - i; j < v.size(); j += p) {
            while (v[j] % p == 0)
                v[j] /= p;
        }
    }
    cout << ans << endl;
    return 0;
}

File: code/291/solution.go
package main /*{{{*/

import "fmt"

const (
	e4  = int64(10000)
	e5  = int64(100000)
	e6  = int64(1000000)
	e7  = int64(10000000)
	e8  = e4 * e4
	e9  = e8 * 10
	mod = int64(1e9 + 7)
)

/*}}}*/

const upper = int64(5e15)

var v = []int64{0}

func main() {
	for n := int64(1); ; n++ {
		now := 2*n*n + 2*n + 1
		if now > upper {
			break
		}
		v = append(v, now)
	}

	ans := 0
	l := int64(len(v))
	for i := int64(1); i < l; i++ {
		p := v[i]
		if p == 2*i*i+2*i+1 {
			ans++
		}
		if p <= 1 {
			continue
		}
		for j := i; j < l; j += p {
			for v[j]%p == 0 {
				v[j] /= p
			}
		}
		for j := p - 1 - i; j < l; j += p {
			for v[j]%p == 0 {
				v[j] /= p
			}
		}
	}
	fmt.Println(ans)
}
